### ISSUE TITLE
TypeError while logging to file with --debug

### ISSUE THREAD

**Issue by inclementweather** [ultrabug/py3status#492]
Just a minor issue while logging debug info to a file I noticed this:

> 2016-10-05 19:07:50 WARNING Instance `netdata`, user method `net_speed` failed (TypeError) core.py line 497.

```
TypeError: write() argument must be str, not bytes
  File "/usr/lib/python3.5/site-packages/py3status/module.py", line 496, in run
    'method {} returned {} '.format(meth, result))
  File "/usr/lib/python3.5/site-packages/py3status/core.py", line 497, in log
    f.write(out.encode('utf-8'))
```

The unicode arrows used by the netdata module are causing the issue. I found adding `encoding='utf8'` to the open() in core.py log method fixes this issue for python3 if you don't mind the unprintable chars in your file.  Python2 needs io.open() to support the encoding parameter though. Since both versions can use io.open() I think it might the best way to avoid TypeError and UnicodeEncodeError for modules like this.

As a side note I'm not sure how good or bad writing utf8 to a file is since open() uses the filesystem encoding setting as the default.  Its different than the default encoding on my system at least.
```
>>> sys.getdefaultencoding()
'utf-8'
>>> sys.getfilesystemencoding()
'ascii'
```
So instead of encoding='utf8' in io.open() if you just use errors='backslashreplace' you will have nice ascii printable unicode escape codes in your file.

If you think the solution is good I can create another pull request.

**Comment by tobes** [comment]
Thanks for this report.

Unfortunately I can't reproduce this error (looks like my filesystem encoding is utf-8).  Looking briefly at the netdata module it could do with some love.  The format is currently fixed (and not very python2 friendly) and it doesn't use py3.safe_format() which tries to help with the unicode issues (though it might fail in some cases).  It should also find the first nic if `eth0` does not exist (like on my system).

Do you have other modules that are outputting unicode? are they working correctly? If they are then maybe just fixing the module is the way to go.  However anything that makes py3status more robust is encouraged.

As a total aside I thing the debug needs cleaning up as it creates so much noise in the log file.  maybe something that allows specific modules or just the core system to be debugged rather than everything might be nice.

**Comment by inclementweather** [comment]
Maybe filter module output if there's no debug parameter set in the module config?  

I don't have any other modules running that are using unicode characters.  I'll go add some in and see what happens.  I'm curious now how these encoding settings are determined, my system's locale is set to en_US.UTF-8.  

Anyway, I have an alternative that deals with both byte and unicode strings without the io import. Its a small change but hopefully it will prevent this issue in the future. I put together a little test script to see if it works on your system too since your encoding settings are different. Would you try it out? Run it through python2 and 3. Let me know if all 8 samples write to the file.  

[test2.py.txt](https://github.com/ultrabug/py3status/files/515458/test2.py.txt)

My output:
```
2.7.12 (default, Jun 28 2016, 08:31:05) 
[GCC 6.1.1 20160602]
FS enc:ANSI_X3.4-1968
DE enc:ascii
Exception:'ascii' codec can't decode byte 0xe2 in position 9: ordinal not in range(128)
sample-1 â â â
sample-2 string literal
sample-3 \u2193 \u2191 \u2195
sample-4 â â â
sample-5 unicode literal
sample-6 â â â
sample-7 byte literal
sample-8 \u2193 \u2191 \u2195
3.5.2 (default, Jun 28 2016, 08:46:01) 
[GCC 6.1.1 20160602]
FS enc:ascii
DE enc:utf-8
sample-1 â â â
sample-2 string literal
sample-3 â â â
sample-4 â â â
sample-5 unicode literal
sample-6 â â â
Exception:'bytes' object has no attribute 'encode'
sample-7 byte literal
Exception:'bytes' object has no attribute 'encode'
sample-8 \u2193 \u2191 \u2195
```

**Comment by tobes** [comment]
My output (looks identical to yours from what I see)

```
2.7.11 (default, Sep  2 2016, 14:51:16) 
[GCC 5.3.1 20160406 (Red Hat 5.3.1-6)]
FS enc:UTF-8
DE enc:ascii
Exception:'ascii' codec can't decode byte 0xe2 in position 9: ordinal not in range(128)
sample-1 ↓ ↑ ↕
sample-2 string literal
sample-3 \u2193 \u2191 \u2195
sample-4 ↓ ↑ ↕
sample-5 unicode literal
sample-6 ↓ ↑ ↕
sample-7 byte literal
sample-8 \u2193 \u2191 \u2195
3.4.3 (default, Aug  9 2016, 15:36:17) 
[GCC 5.3.1 20160406 (Red Hat 5.3.1-6)]
FS enc:utf-8
DE enc:utf-8
sample-1 ↓ ↑ ↕
sample-2 string literal
sample-3 ↓ ↑ ↕
sample-4 ↓ ↑ ↕
sample-5 unicode literal
sample-6 ↓ ↑ ↕
Exception:'bytes' object has no attribute 'encode'
sample-7 byte literal
Exception:'bytes' object has no attribute 'encode'
sample-8 \u2193 \u2191 \u2195
```

> Maybe filter module output if there's no debug parameter set in the module config? 

That's an option but I'd rather specify the debugging through the command line args rather than make changes to the config.

**Comment by inclementweather** [comment]
That's great, thanks.  I think that sample set covers anything that would be passed to the logger.

**Comment by tobes** [comment]
>  I think that sample set covers anything that would be passed to the logger. 

In theory we can also log non-string types eg `True`, `None`, `{'key': value}`, `['my', 'list']`, `MyObject()` but these should be handled ok if the basics are working.  This is useful when debugging etc.

Logging can go through two paths `core.Py3statusWrapper.log()` and `py3.Py3.log()` which is a wrapper for use in modules.  Looking at the code I think that `py3.py3.log()` should probably be returning unicode rather than str under python2, but we need to be able to deal with whatever gets thrown at us.

As well as the custom logfile, logging can go to syslog has that been working ok for you?

**Comment by inclementweather** [comment]
Yeah the logging to syslog is working fine, but I'm on Arch so systemd is writing to a journal and I don't have a text based logging daemon set up. 

I think you're right about py3.py3.log(). Everything I read while looking at this unicode business said that converting incoming data to unicode as soon as possible is the best practice.  core.Py3statusWrapper.log() is converting msg to unicode anyway but it seems wise not to rely on such implicit functions. 

I just tried to pass some non-string types and it definitely doesn't like those. How should it handle types like these?  What kind of info should be written to the log?

**Comment by tobes** [comment]
> I just tried to pass some non-string types and it definitely doesn't like those.

When you say this are you talking about your `test2.py` script or in py3status?  As far as I'm aware in py3status at least for `py3.log()` this should work fine and it is something I use regularly.

> Also I tried adding a unicode char to the sysdata module through the config and it seems the parser eats the escape so just the u3184 shows up. If I modify the default status line in the source though it causes the same issue. 

This is two issues.  If you add the `\u3184` to the module default it needs to be as unicode eg `format = u'\u3184 ...'` if you are running using python2 for me at least, it works as expected using python3.  The config parser does have an issue (with both python 2/3) where it is fine with explicit unicode eg `↓` but not things like `\u3184` (I've created a fix for this)

**Comment by inclementweather** [comment]
Yeah, the test script. I just passed them bare into the write function out of curiosity. As soon as you pass it to format though it becomes text so those types would never reach that point. Forget my ignorant question above. :)

Nice fix :+1:  I'll have to add some more unicode to my status bar now.  

The unicode (explicit or escaped) in the default format shows up fine in the status bar, but it was causing the same error when logging to a file.  Only an issue if someone had FS encoding set to ascii.

**Comment by tobes** [comment]
> but it was causing the same error when logging to a file. Only an issue if someone had FS encoding set to ascii.

Well I thought that was something you were fixing ;)  The config parsing is a separate issue which is why I sorted it.

**Comment by inclementweather** [comment]
I'm on it boss :) I was actually starting to do that, but I got sucked back into this python2 unicode maze. I set up a virtualenv to run py3status in python2 because I expected the unicode to cause more errors when it came to format(), but it doesn't. It seems including a byte string in a dict (maybe even a list) converts the unicode characters to their escaped byte equivalents, and then passes through the log function no problem, even with my fs encoding set to ascii. You can also format a unicode string and dict containing byte string with unicode characters and it doesn't try to decode it as it would if it was passed not in a dict. This may be why python2 wasn't having issues.

```
Python 2.7.12 (default, Jun 28 2016, 08:31:05)
[GCC 6.1.1 20160602] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> test = "sample-1 ↓ ↑ ↕"
>>> testdict = { "sample1": "sample-1 ↓ ↑ ↕" }
>>> u"{}".format(test)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 9: ordinal not in range(128)
>>> u"{}".format(testdict)
u"{'sample1': 'sample-1 \\xe2\\x86\\x93 \\xe2\\x86\\x91 \\xe2\\x86\\x95'}"

```
Double escaped I think by the interpreter? Not sure because it looks like this in the log file:

`2016-10-08 22:15:40 INFO method net_speed returned {'color': u'#FFFFFF', 'instance': '', 'name': u'netdata', 'full_text': 'LAN(Kb):   0.5\xe2\x86\x93   0.1\xe2\x86\x91'} `

You can't do it in python3 because it won't let you create a byte string with unicode characters.

I tested the changes out every which way I can think of, python2 + python3 with filesystem encoding set to utf-8 and ascii, unicode characters in the default format or in the config file, and escaped or raw unicode in the default format. 

The only thing to fail was Python2.  If you add raw unicode to the config format line you get the error below.

```
2016-10-08 22:24:38 INFO Traceback
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc2 in position 4: ordinal not in range(128)
  File "/home/tx/.virtualenvs/py2env/lib/python2.7/site-packages/py3status/module.py", line 419, in run
    response = method()
  File "/home/tx/.virtualenvs/py2env/lib/python2.7/site-packages/py3status/modules/sysdata.py", line 153, in sysData
    mem_used_percent='%.2f' % mem_used_percent,
```

Everything else works fine though. I'll set up a pull request.

**Comment by tobes** [comment]
fixed in #490
